.NET Core csc output non-ASCII characters correctly in the Console output

This issue is spawned from: https://github.com/dotnet/cli/issues/2486, which is an RC2 partner blocking bug (VS Code).

**Version Used**: 1.3.0.60410

**Steps to Reproduce**:
1. I have an .rsp file with the following input file:

```
"C:\Users\eerhardt\Tést\Program.cs"
```

And that Program.cs file contains a build error.
1. Compile it using the .NET Core csc
   - C:\cli\artifacts\win10-x64\stage1\dotnet.exe C:\cli\artifacts\win10-x64\stage1\sdk\1.0.0-rc2-002476\csc.dll @C:\Temp\dotnet-compile-csc.rsp

**OLD REPRO:**
    cd /d %USERPROFILE%
    mkdir Tést
    cd Tést
    dotnet new
    dotnet restore
    dotnet build

**Expected Behavior**:

The strings written to the console should contain the correct path:

```
C:\Users\eerhardt\Tést\Program.cs(9,17): error CS1001: Identifier expected
C:\Users\eerhardt\Tést\Program.cs(9,17): error CS1002: ; expected
```

**Actual Behavior**:

Incorrect strings are written to the console:

```
C:\Users\eerhardt\T├⌐st\Program.cs(9,17): error CS1001: Identifier expected
C:\Users\eerhardt\T├⌐st\Program.cs(9,17): error CS1002: ; expected
```

**OLD behavior:**
On my machine I'm seeing a crash in the csc process:

```
Unhandled Exception: System.IO.FileNotFoundException: C:\cli\artifacts\win10-x64\stage1\sdk\1.0.0-rc2-002476\Microsoft.CodeAnalysis.dll
   at System.Diagnostics.FileVersionInfo.GetVersionInfo(String fileName)
   at Roslyn.Utilities.PortableShim.Misc.GetFileVersion(String path)
   at Microsoft.CodeAnalysis.CommonCompiler.GetAssemblyFileVersion()
   at Microsoft.CodeAnalysis.CSharp.CSharpCompiler.PrintLogo(TextWriter consoleOutput)
   at Microsoft.CodeAnalysis.CommonCompiler.RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CommonCompiler.Run(TextWriter consoleOutput, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CommandLine.Csc.<>c__DisplayClass1_0.<Run>b__0(TextWriter tw)
   at Microsoft.CodeAnalysis.CSharp.CommandLine.Csc.Run(String[] args, BuildPaths buildPaths, TextWriter textWriter, IAnalyzerAssemblyLoader analyzerLoader)
   at Microsoft.CodeAnalysis.CommandLine.CoreClrBuildClient.RunLocalCompilation(String[] arguments, BuildPaths buildPaths, TextWriter textWriter)
   at Microsoft.CodeAnalysis.CommandLine.BuildClient.RunCompilation(IEnumerable`1 originalArguments, BuildPaths buildPaths, TextWriter textWriter)
   at Microsoft.CodeAnalysis.CommandLine.CoreClrBuildClient.Run(IEnumerable`1 arguments, RequestLanguage language, CompileFunc compileFunc)
   at Microsoft.CodeAnalysis.CSharp.CommandLine.Program.Main(String[] args)
```

I can pare the "dotnet build" call down to starting the following process:

```
C:\Users\eerhardt\Tést>C:\cli\artifacts\win10-x64\stage1\dotnet.exe C:\cli\artifacts\win10-x64\stage1\sdk\1.0.0-rc2-002476\csc.dll -noconfig @C:\Users\eerhardt\Tést\obj\Debug\netcoreapp1.0\dotnet-compile-csc.rsp
```

Looking through the callstack, it is trying to write the logo, so I shut that off:

```
C:\Users\eerhardt\Tést>C:\cli\artifacts\win10-x64\stage1\dotnet.exe C:\cli\artifacts\win10-x64\stage1\sdk\1.0.0-rc2-002476\csc.dll -nologo -noconfig @C:\Users\eerhardt\Tést\obj\Debug\netcoreapp1.0\dotnet-compile-csc.rsp
```

and now I get:

```
error CS2011: Error opening response file 'C:\Users\eerhardt\T├â┬⌐st\obj\Debug\netcoreapp1.0\dotnet-compile-csc.rsp'
warning CS2008: No source files specified.
error CS1562: Outputs without source must have the /out option specified
```
# Solution

The underlying problem is a difference in encoding handling in .NET Core vs. .NET Framework. (See the ".NET Core Notes" section in https://msdn.microsoft.com/en-us/library/system.diagnostics.process(v=vs.110).aspx) .NET Core uses a minimal set of code pages, and falls back to Utf8 when it can't find the right code page. In order to get the same behavior as the .NET Framework, console apps have to add the correct code pages into their process.

One possible way of doing this is:

``` C#
using System.Text;
...
    Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
```

~~In order to fix https://github.com/dotnet/cli/issues/2486, we need to make that same call in the CLI's Main method.  But to fully fix the scenario, we need Roslyn to also add this call in the .NET Core version of csc.exe's Main method.~~

/cc @Petermarcu @jaredpar 
/bcc @tarekgh - who provided me guidance on the .NET Core behavior difference

@agocke's comment is correct.  The blocking issue I ran into is actually a dotnet.exe corehost issue not passing in the command line parameters correctly to the managed code.

Re-naming and re-purposing (and lowering the priority) this issue for Roslyn to handle non-ASCII characters in their Console output correctly.

To quote from https://github.com/dotnet/corefx/issues/7935#issuecomment-212691693

> Also the control of the console encoding can be done either from the system settings or from the console process itself (through Console.InputEncoding and Console.OutputEncoding). 

To repro:

I have an .rsp file with the following input file:

```
"C:\Users\eerhardt\Tést\Program.cs"
```

And that Program.cs file contains a build error.

I compile it using the desktop .NET Framework's csc.exe and I get the following output written to the console:

```
C:\DotNetTest\ConsoleApp2>csc.exe @C:\Temp\dotnet-compile-csc.rsp
C:\Users\eerhardt\Tést\Program.cs(9,17): error CS1001: Identifier expected
C:\Users\eerhardt\Tést\Program.cs(9,17): error CS1002: ; expected
```

But when I compile it using the .NET Core's csc, I don't get `Tést` written correctly:

```
C:\DotNetTest\ConsoleApp2>C:\cli\artifacts\win10-x64\stage1\dotnet.exe C:\cli\artifacts\win10-x64\stage1\sdk\1.0.0-rc2-002476\csc.dll @C:\Temp\dotnet-compile-csc.rsp
C:\Users\eerhardt\T├⌐st\Program.cs(9,17): error CS1001: Identifier expected
C:\Users\eerhardt\T├⌐st\Program.cs(9,17): error CS1002: ; expected
```

--- a/src/Compilers/Shared/CoreClrBuildClient.cs
+++ b/src/Compilers/Shared/CoreClrBuildClient.cs
@@ -27,6 +27,7 @@ namespace Microsoft.CodeAnalysis.CommandLine
         internal static int Run(IEnumerable<string> arguments, RequestLanguage language, CompileFunc compileFunc)
         {
             // Register encodings for console
+            // https://github.com/dotnet/roslyn/issues/10785 
             Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
 
             // Should be using BuildClient.GetCommandLineArgs(arguments) here.  But the native invoke

--- a/src/Compilers/Shared/CoreClrBuildClient.cs
+++ b/src/Compilers/Shared/CoreClrBuildClient.cs
@@ -6,6 +6,7 @@ using System.Collections.Generic;
 using System.Net.Sockets;
 using System.Threading;
 using System.Threading.Tasks;
+using System.Text;
 
 namespace Microsoft.CodeAnalysis.CommandLine
 {
@@ -25,6 +26,9 @@ namespace Microsoft.CodeAnalysis.CommandLine
 
         internal static int Run(IEnumerable<string> arguments, RequestLanguage language, CompileFunc compileFunc)
         {
+            // Register encodings for console
+            Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
+
             // Should be using BuildClient.GetCommandLineArgs(arguments) here.  But the native invoke
             // ends up giving us both CoreRun and the exe file.  Need to find a good way to remove the host
             // as well as the EXE argument.

--- a/src/Compilers/Shared/CoreClrBuildClient.cs
+++ b/src/Compilers/Shared/CoreClrBuildClient.cs
@@ -1,12 +1,12 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
-using System.IO;
 using System.Collections.Generic;
+using System.IO;
 using System.Net.Sockets;
+using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
-using System.Text;
 
 namespace Microsoft.CodeAnalysis.CommandLine
 {

